source/MaterialXGenShader/Nodes/LayerNode.cpp

```diff
@@ -0,0 +1,90 @@
+//
+// TM & (c) 2017 Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved.  See LICENSE.txt for license.
+//
+
+#include <MaterialXGenShader/Nodes/LayerNode.h>
+#include <MaterialXGenShader/GenContext.h>
+#include <MaterialXGenShader/ShaderNode.h>
+#include <MaterialXGenShader/ShaderStage.h>
+#include <MaterialXGenShader/ShaderGenerator.h>
+#include <MaterialXGenShader/TypeDesc.h>
+
+namespace MaterialX
+{
+
+const string LayerNode::TOP_STRING = "top";
+const string LayerNode::BASE_STRING = "base";
+
+ShaderNodeImplPtr LayerNode::create()
+{
+    return std::make_shared<LayerNode>();
+}
+
+void LayerNode::emitFunctionCall(const ShaderNode& _node, GenContext& context, ShaderStage& stage) const
+{
+    BEGIN_SHADER_STAGE(stage, Stage::PIXEL)
+        const ShaderGenerator& shadergen = context.getShaderGenerator();
+
+        ShaderNode& node = const_cast<ShaderNode&>(_node);
+
+        ShaderInput* top = node.getInput(TOP_STRING);
+        ShaderInput* base = node.getInput(BASE_STRING);
+        ShaderOutput* output = node.getOutput();
+        if (!(top && base && output))
+        {
+            throw ExceptionShaderGenError("Node '" + node.getName() + "' is not a valid layer node");
+        }
+        if (!top->getConnection())
+        {
+            // No top layer, just emit default value.
+            shadergen.emitLineBegin(stage);
+            shadergen.emitOutput(output, true, true, context, stage);
+            shadergen.emitLineEnd(stage);
+            shadergen.emitLineEnd(stage);
+            return;
+        }
+        
```

Comment:
Minor: 
- Should this implementation and corresponding nodedef be called something more specific than "layer". Unsure if this will this become a general purpose node for layering any type? 
- Should finding a base connection throw an exception ?

Reply:
The proposed names we have discussed earlier with Lucasfilm is "layer" or "stack". I like "layer" the best but we can discuss more with Lucasfilm when we merge this to master.

When/if we decide that using layer operators is the recommended method for layering we can enforce this by throwing exception, or hiding the base input.